filecheck/src/FileCheck.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module FileCheck
+  ( Output(..)
+  , FCD.Directive(..)
+  , FCC.Command(..)
+  , FileCheckFailure(..)
+  , check
+  , check'
+  , CommandParseFailure
+  , parseCommentsAndCheck
+  , parseCommentsAndCheck'
+  ) where
+
+import Control.Exception qualified as X
+import Control.Monad qualified as Monad
+import Data.Maybe qualified as Maybe
+import Data.Text qualified as Text
+import Data.Text (Text)
+import FileCheck.Command (Command)
+import FileCheck.Command qualified as FCC
+import FileCheck.Directive qualified as FCD
+import FileCheck.Directive (Prefix)
+import GHC.Stack (HasCallStack)
+import Prelude hiding (lines)
+
+-- | Output of the program under test
+newtype Output = Output Text
+
+data FileCheckFailure = FileCheckFailure Command Text
+
+instance Show FileCheckFailure where
+  show (FileCheckFailure (FCC.Command d c) t) =
+    Text.unpack $
+      Text.unlines
+      [ FCD.print d <> ": " <> c
+      , "does not match"
+      , t
+      ]
+
+instance X.Exception FileCheckFailure
+
+-- | Match text against a sequence of 'Command's.
+check ::
+  HasCallStack =>
+  [Command] ->
+  Output ->
+  Either FileCheckFailure ()
```

Comment:
This function doesn't throw any exceptions. Does it really need a `HasCallStack` constraint?

Reply:
Nope! Good call.